[CMake] Get SwiftCore building with the new build system

etcwilde opened the pull request:
This PR gets us up to SwiftCore building with the new build system.
I've wired it into the existing build and am enabling building with it in the Linux smoke-tests to keep things going. We are not currently running or installing the built code though.

rdar://139221398

etcwilde committed:
Get swiftLLVMSupport building

This patch gets the LLVM Support object files building.

etcwilde committed:
Add SwiftShims as an interface library

etcwilde committed:
Add SwiftDemangling

etcwilde committed:
Build threading library

etcwilde committed:
Build runtime library

Getting the runtime libraries building, while also ironing out more of
the macro definitions and flags.

etcwilde committed:
Add stubs library to build

Adding stdlib stubs library, which uses gyb and requires knowing about
some of the platform info.

The stubs library pulls in several headers from the compiler repository.
We should probably clean that up, but not right now. I've made a note of
it.

etcwilde committed:
Add swiftCore library

This patch hooks up the swiftCore library build and gets it installing.
This means that we have library evolution hooked up and vector types.

Building it dynamically found that we had duplicate symbols, so fixed
the swiftDemanglingCR library.

Needed to hook up the availability macros.

The flag configuration is for macOS, so we'll need to go through and
figure out what it should look like for the other platforms too.

etcwilde committed:
Setup reasonable defaults

This patch makes it simpler to setup reasonable defaults for a given
platform. The standard library has many knobs for configuration
purposes. This is great for providing cache files to configure specific
builds, but the build system should still generally work if someone runs
a minimal CMake invocation against the build without futzing with
various options.

etcwilde committed:
Hook new build into existing Swift project

Setting `SWIFT_ENABLE_NEW_RUNTIME_BUILD` to `ON` will build and run the
new standard library build in addition to using the old build. This is
primarily for basic smoke testing at the moment. Once we're confident,
we can start running tests against the stdlib built with the new
build system.

The mechanism will build the stdlib for each of the same platforms that
the old build system would build for, but emitting a separate build
directory with a separate CMake invocation for each configuration.

This will use the just-built compiler if the compiler was not
cross-compiled. If the compiler were cross-compiled, it will use the
compiler from the toolchain that is known to work on the builder.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
Linux test failure:

```
CMake Error at CMakeLists.txt:32 (cmake_minimum_required):
  CMake 3.26 or higher is required.  You are running version 3.24.2
```

etcwilde commented:
We'll need to update our CMake version

mikeash commented on Runtimes/Core/CMakeLists.txt:
```diff
@@ -0,0 +1,83 @@
+# functionality. When building the demangling library, the macro should be set
+# to 0 or 1 to indicate the presence of the crashreporter.
+# When building the runtime library, the existence of the macro indicates the
+# presence of the crashreporter.
```
Can we set it to 1 when enabled and un-set it when disabled? That should work for both.

etcwilde replied:
As far as I can tell, C/C++ defines it so that if a `-DFOO` is passed without a value, it's given the value of `1`, and `#if FOO` will see undefined an `0` the same way. We might be able to get away with just setting it. I'll have to compare the symbol lists though.

mikeash commented on Runtimes/Core/cmake/modules/CompilerChecks/ptr_size:
```diff
@@ -0,0 +1 @@
+__SIZEOF_POINTER__
```
Should this be gitignored? Looks like `PlatformInfo.cmake` writes this when it's needed. Or maybe `PlatformInfo.cmake` doesn't need to write it.

mikeash replied:
Right, but it's also created there:

```
  file(WRITE ${CMAKE_BINARY_DIR}/cmake/CompilerChecks/ptr_size "__SIZEOF_POINTER__")
  set(size_check_command ${CMAKE_C_COMPILER} -x c -P -E -o - ${CMAKE_BINARY_DIR}/cmake/CompilerChecks/ptr_size)
```

I'm thinking it should be committed, or written when needed, but probably not both.

etcwilde replied:
Yeah, this file isn't necessary. Since we have C enabled, I can probably just use `CMAKE_SIZEOF_VOID_P` and skip this. I was having some issues with it before where it was empty/wrong with older versions of CMake, but it looks like it's behaving correctly now.

drodriguez commented on Runtimes/Core/cmake/modules/PlatformInfo.cmake:
```diff
@@ -0,0 +1,51 @@
+if(NOT SwiftCore_SWIFTC_CLANGIMPORTER_RESOURCE_DIR)
+  # TODO: We need to separate the concept of compiler resources and the stdlib.
+  #       Compiler-resources in the compiler-resource directory are specific to
+  #       a given compiler. The headers in `lib/clang/include` and
```
Nit: I might be thinking of other directory, but I think it is normally `lib/clang/<version>/include`.

drodriguez commented on Runtimes/Core/cmake/modules/gyb.cmake:
```diff
@@ -0,0 +1,47 @@
+  endforeach()
+  list(REMOVE_DUPLICATES gyb_expand_deps)
+
+  set(utils_dir "${PROJECT_SOURCE_DIR}/../../utils/")
```
I am seeing this `${PROJECT_SOURCE_DIR}/../..` bit in several places. I don't think `Runtimes/Core` has a lot of chances of changing directory levels, but in case it does, chasing all these might not be fun. I was thinking in having a `SWIFT_TOP_LEVEL_SOURCE_DIR` calculated once after the `project()` call, which can be used in the rest of the files. It also happens to document how to get to the top level source dir, if necessary.

drodriguez commented on Runtimes/Core/core/CMakeLists.txt:
```diff
@@ -0,0 +1,341 @@
+  FLAGS "-DCMAKE_SIZEOF_VOID_P=${SwiftCore_SIZEOF_POINTER}")
+
+# These sources are not strictly sorted alphabetically because the compiler
+# crashes if it is.
```
Nit: `# crashes if they are` (if we are talking about "sources", right?)

compnerd commented on Runtimes/Core/core/CMakeLists.txt:
```diff
@@ -0,0 +1,341 @@
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xfrontend -require-explicit-availability=ignore>"
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xfrontend -enforce-exclusivity=unchecked>"
+  "$<$<COMPILE_LANGUAGE:Swift>:-disable-autolinking-runtime-compatibility-dynamic-replacements>"
+  "$<$<COMPILE_LANGUAGE:Swift>:-no-link-objc-runtime>"
```
Can we hoist this to the initial set of options?

compnerd commented on Runtimes/Core/core/CMakeLists.txt:
```diff
@@ -0,0 +1,341 @@
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xfrontend -enforce-exclusivity=unchecked>"
+  "$<$<COMPILE_LANGUAGE:Swift>:-disable-autolinking-runtime-compatibility-dynamic-replacements>"
+  "$<$<COMPILE_LANGUAGE:Swift>:-no-link-objc-runtime>"
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-library-level api>"
```
Can we hoist this before the `-Xfrontend` group?

compnerd commented on Runtimes/Core/core/CMakeLists.txt:
```diff
@@ -0,0 +1,341 @@
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xfrontend -disable-implicit-string-processing-module-import>"
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xfrontend -prespecialize-generic-metadata>"
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-runtime-compatibility-version none>"
+  "$<$<COMPILE_LANGUAGE:Swift>:-warn-implicit-overrides>"
```
Can we hoist this to the initial set of options?

compnerd commented on Runtimes/Core/core/CMakeLists.txt:
```diff
@@ -0,0 +1,341 @@
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xfrontend -disable-implicit-backtracing-module-import>"
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xfrontend -disable-implicit-string-processing-module-import>"
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xfrontend -prespecialize-generic-metadata>"
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-runtime-compatibility-version none>"
```
Can we hoist this before the `-Xfrontend` options?

compnerd commented on Runtimes/Core/runtime/CMakeLists.txt:
```diff
@@ -75,6 +80,7 @@ target_compile_definitions(swiftRuntime
     -DSWIFT_TARGET_LIBRARY_NAME=swiftRuntime
     -DswiftCore_EXPORTS
     $<$<BOOL:${SwiftCore_ENABLE_REFLECTION}>:-DSWIFT_ENABLE_REFLECTION>
+    -DSWIFT_LIBRARY_EVOLUTION=$<BOOL:${SwiftCore_ENABLE_LIBRARY_EVOLUTION}>
```
Hoist this before the optional flags?

compnerd commented on CMakeLists.txt:
```diff
@@ -1579,3 +1579,45 @@ if(XCODE)
+  endforeach()
+
+  # Materialize sources in the new build
+  execute_process(COMMAND
```
This only runs once when you configure, and won't run again. Should we do this as a prebuild command or something?

drodriguez replied:
The readme had a warning about this:

https://github.com/swiftlang/swift/blob/main/Runtimes/Readme.md?plain=1#L15-L17

I don't know if there's a reason to not do that process automatically, in case someone wants to update their sources of the compiler, but not incorporate changes from the runtimes  (or why it needs to happen automatically during configuring, besides CI testing).

drodriguez commented on Runtimes/Core/Demangling/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+  "${PROJECT_SOURCE_DIR}/../../lib/Demangling/Punycode.cpp"
+  "${PROJECT_SOURCE_DIR}/../../lib/Demangling/Remangler.cpp"
+  "${PROJECT_SOURCE_DIR}/../../lib/Demangling/NodeDumper.cpp"
+  "${PROJECT_SOURCE_DIR}/../../lib/Demangling/Errors.cpp")
```
I like this idea better than mine of calculating `${PROJECT_SOURCE_DIR}/../..` only once in the project `CMakeLists.txt`.

drodriguez commented on Runtimes/Core/Demangling/CMakeLists.txt:
```diff
@@ -9,27 +9,26 @@ add_library(swiftDemangling STATIC
+target_compile_definitions(swiftDemangling PRIVATE swiftCore_EXPORTS
+  -DSWIFT_SUPPORT_OLD_MANGLING=$<BOOL:${SwiftCore_ENABLE_OBJC_INTEROP}>
+  $<$<BOOL:${SwiftCore_ENABLE_TYPE_PRINTING}>:-DSWIFT_STDLIB_HAS_TYPE_PRINTING>
+  -DSWIFT_HAVE_CRASHREPORTERCLIENT=$<BOOL:${SwiftCore_ENABLE_CRASH_REPORTER_CLIENT}>)
```
I looked for this in the docs when I was reviewing the changes. They are allowed. Not required. I don't have a strong opinion about having or not having them.

etcwilde replied:
I prefer the `-D` prefixes. It's just a style thing.

etcwilde commented on Runtimes/Core/cmake/modules/PlatformInfo.cmake:
```diff
@@ -0,0 +1,51 @@
+if(NOT SwiftCore_SIZEOF_POINTER)
+  # This seems to be the most reliable way to actually get the size of a pointer
```
`CMAKE_SIZEOF_VOID_P` is probably better. It wasn't working for me when I first tried it, but looks like it is now.

etcwilde commented on Runtimes/Core/core/CMakeLists.txt:
```diff
@@ -0,0 +1,341 @@
+  UnicodeHelpers.swift
+  UnicodeParser.swift
+  UnicodeScalarProperties.swift
+  CharacterProperties.swift # ORDER DEPENDENCY: UnicodeScalarProperties.swift
```
Why? It comes after `UnicodeScalarProperties.swift` (and doesn't crash).

compnerd replied:
oh ... I thought it was supposed to be before!

etcwilde commented on Runtimes/Core/runtime/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+
+target_include_directories(swiftRuntime PRIVATE
+  "${PROJECT_BINARY_DIR}/include"
+  ./)
```
I only want the current source directory, not the current binary directory, and I only want it on the `swiftRuntime` target.

I could use a `block()`, but that seems excessive. `${CMAKE_CURRENT_SOURCE_DIR}` is fine, but CMake swaps out `.` for `${CMAKE_CURRENT_SOURCE_DIR}` in this context anyway.

compnerd replied:
I'm less concerned about the swapping of the value and more about the readability. I feel like `.` on its own is easy to miss.

etcwilde committed:
Moving compiler-src-dir refs

Moving the rest of the references to the compiler source directory to
use the `SwiftCore_SWIFTC_SOURCE_DIR` variable instead of
`${PROJECT_SOURCE_DIR}/../../`.

etcwilde committed:
Fixing comments

Fixing some of the comments in DefaultSettings.cmake and
PlatformInfo.cmake.

I haven't added any vendor cache files yet, so the comment doesn't point
anywhere. Also fixing the reference to the location of the clang
resource headers in PlatformInfo.cmake.

etcwilde committed:
Get Android building

etcwilde committed:
Auto populate runtime source directory

This patch causes the convenient flow to automatically populate the
sources into the new runtime directory layout. This is run at build-time
so that it repopulates on each build. Since the resync script uses
`COPY_IF_DIFFERENT`, only files that change in the main source directory
are copied over, so incremental builds should still work.

Fixed a small bug in the resync script to ensure that it creates the
subdirectories in the right places.

etcwilde committed:
Fix rpath flag

etcwilde commented on Runtimes/Core/cmake/modules/PlatformInfo.cmake:
```diff
@@ -0,0 +1,51 @@
+  #       the one copy of clang should never be looking in the other's resource
+  #       directory. If there are issues here, something has gone horribly wrong
+  #       and you're looking in the wrong place.
+  set(module_triple_command "${CMAKE_Swift_COMPILER}" -print-target-info)
+  if(CMAKE_Swift_COMPILER_TARGET)
+    list(APPEND module_triple_command -target ${CMAKE_Swift_COMPILER_TARGET})
+  endif()
+  execute_process(COMMAND ${module_triple_command} OUTPUT_VARIABLE target_info_json)
```
CMake doesn't allow newlines in variables written to the cache so storing the JSON emitted by the driver won't work.
The commands should only get hit on the initial cmake invocation since `SwiftCore_SWIFTC_CLANGIMPORTER_RESOURCE_DIR` and `SwiftCore_MODULE_TRIPLE` will be set and cached though.

If it becomes a noticeable issue, we can come back to this though.

compnerd replied:
Sorry, poor choice of words on my part. I didn't mean into `$CACHE` but rather into a local variable so that we don't continually invoke `swiftc -print-target-info -target x86_64-unknown-windows-msvc`.

etcwilde commented on Runtimes/Core/runtime/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+  target_link_libraries(swiftrt PRIVATE swiftShims)
+  install(TARGETS swiftrt DESTINATION "${CMAKE_INSTALL_LIBDIR}/swift")
+elseif(NOT "${SwiftCore_OBJECT_FORMAT}" STREQUAL "x")
+  message(SEND_ERROR "Unknown object format '${SwiftCore_OBJECT_FORMAT}'")
+endif()
```
The swiftrt target doesn't always exist.

compnerd replied:
Oh, in the MachO case 🤦‍♂️

etcwilde commented on Runtimes/Core/cmake/modules/CMakeWorkarounds.cmake:
```diff
@@ -0,0 +1,31 @@
+# Swift compiler.
+
+# FIXME: This is a workaround for the broken Swift compiler that cannot
+#        typecheck the Swift stdlib without crashing. <ADD RADAR HERE>.
```
Yeah, not sure what I want out of the bug report. Do I want the compiler fixed? Do I want CMake to have support for this? ¯\_(ツ)_/¯ I don't know!

compnerd replied:
I would guess you want the compiler fixed, but I don't know what you were thinking initially.

etcwilde replied:
I mostly just want this file to not have to exist. I guess I can reword this and just make it an explanation about why I'm doing this.

etcwilde committed:
Shuffle definitions around

De-duping some of the definitions. We can pass most of the definitions
to both the Swift and C/C++ compilers in the same form if we don't
assign anything. C/C++ will set the macro value to `1` if there is no
`=` as part of the definition, and `#if` recognizes a non-existent macro
to be false. With this logic, we can unify some of these.

etcwilde commented:
@swift-ci please test

compnerd commented on Runtimes/Core/cmake/modules/CMakeWorkarounds.cmake:
```diff
@@ -0,0 +1,31 @@
+  set(CMAKE_Swift_ARCHIVE_CREATE "<CMAKE_AR> crs <TARGET> <OBJECTS>")
+  set(CMAKE_Swift_ARCHIVE_FINISH "")
+
+  add_compile_options($<$<COMPILE_LANGUAGE:Swift>:-wmo>)
```
Can we add a comment about that? I feel like I didn't catch that and it is easy to forget that.

compnerd commented on Runtimes/Core/Demangling/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+target_compile_definitions(swiftDemangling PRIVATE swiftCore_EXPORTS)
+
+target_include_directories(swiftDemangling
+  PRIVATE "${PROJECT_SOURCE_DIR}/../../include")
```
Is this not `${SwiftCore_SWIFTC_SOURCE_DIR}/include`?

compnerd commented on Runtimes/Core/Threading/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+  "${PROJECT_SOURCE_DIR}/../../lib/Threading/C11.cpp"
+  "${PROJECT_SOURCE_DIR}/../../lib/Threading/Linux.cpp"
+  "${PROJECT_SOURCE_DIR}/../../lib/Threading/Pthreads.cpp"
+  "${PROJECT_SOURCE_DIR}/../../lib/Threading/Win32.cpp")
```
We can use `SwiftCore_SWIFTC_SOURCE_DIR` here no?

etcwilde replied:
I think it is on the most recent commit?

compnerd commented on Runtimes/Core/Demangling/CMakeLists.txt:
```diff
@@ -9,27 +9,26 @@ add_library(swiftDemangling STATIC
 target_include_directories(swiftDemangling
-  PRIVATE "${PROJECT_SOURCE_DIR}/../../include")
+  PRIVATE
+    "${PROJECT_SOURCE_DIR}/../../include"
```
Is this not `${SwiftCore_SWIFTC_SOURCE_DIR}/include`?

etcwilde replied:
Missed it.

etcwilde committed:
Cleanup WMO forcing

This patch cleans up the handling of WMO-forcing in the workaround file.
`-wmo` is passed in each of the link jobs in the non-CMP0157-enabled
configuration to keep it matching the behavior of the CMP0157-enabled
builds.

This also moves the commentary about num-threads and WMO to the top of
the file to make it clearer what is going on. Ideally the file wouldn't
have to exist, but it does.

The file also contains the target install-name fix that landed in CMake
3.29.8 and CMake 3.30.2, so earlier versions of CMake 3.29 will work
with this build as well.

etcwilde committed:
Remove SwiftCore_SWIFTC_CLANGIMPORTER_RESOURCE_DIR

The SwiftCore_SWIFTC_CLANGIMPORTER_RESOURCE_DIR is not being used so I'm
deleting it for now. This means we only grab the target info once.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please test Linux

etcwilde commented on Runtimes/Core/core/CMakeLists.txt:
```diff
@@ -0,0 +1,341 @@
+
+gyb_expand(Tuple.swift.gyb Tuple.swift
+  FLAGS "-DCMAKE_SIZEOF_VOID_P=${SwiftCore_SIZEOF_POINTER}")
+
```
We could. I like the control over the input/output name and controlling what flags are passed to each gyb call.

etcwilde commented on Runtimes/Core/Demangling/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+# FIXME: Refactor demangling library so that we aren't pulling sources from
```
I'll add it in a separate PR, but here's the issue: https://github.com/swiftlang/swift/issues/77462
